File: lab07/1.py
import numpy as np

# Функция для вычисления правой части уравнения
def f(x, y):
    return -y**2 - 1/x**4

# Точное решение для сравнения
def exact_solution(x):
    return -((x - np.tan((-1 + x)/x)) / x**2)

# Начальные условия
x0, y0 = 1, 1
x_end = 2
h = 0.1
n = int((x_end - x0) / h)

# Массивы для хранения значений
x = np.linspace(x0, x_end, n+1)
y = np.zeros(n+1)
y[0] = y0

# Первый улучшенный метод Эйлера
for i in range(n):
    k1 = f(x[i], y[i])
    k2 = f(x[i] + h, y[i] + h * k1)
    y[i+1] = y[i] + h/2 * (k1 + k2)

# Вычисление точного решения и погрешности
exact_y = exact_solution(x)
errors = np.abs(exact_y - y)

# Вывод результатов
for i in range(n+1):
    print(f"x = {x[i]:.1f}, Numerical y = {y[i]:.6f}, Exact y = {exact_y[i]:.6f}, Error = {errors[i]:.6f}")

File: lab07/1.cpp
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

// Функция для вычисления правой части уравнения
double f(double x, double y) {
    return -y*y - 1/(x*x*x*x);
}

// Точное решение для сравнения
double exact_solution(double x) {
    return -( (x - tan((-1 + x)/x)) / (x*x) );
}

int main() {
    // Начальные условия
    double x0 = 1.0, y0 = 1.0;
    double x_end = 2.0;
    double h = 0.1;
    int n = (int)((x_end - x0) / h);

    // Массивы для хранения значений
    double x[n+1], y[n+1];
    x[0] = x0;
    y[0] = y0;

    // Первый улучшенный метод Эйлера (метод Эйлера с модификацией)
    for (int i = 0; i < n; ++i) {
        x[i+1] = x[i] + h;
        double k1 = f(x[i], y[i]);
        double k2 = f(x[i] + h, y[i] + h * k1);
        y[i+1] = y[i] + h/2 * (k1 + k2);
    }

    // Вывод результатов
    cout << fixed << setprecision(6);
    for (int i = 0; i <= n; ++i) {
        double exact_y = exact_solution(x[i]);
        double error = fabs(exact_y - y[i]);
        cout << "x = " << x[i] << ", Numerical y = " << y[i] << ", Exact y = " << exact_y << ", Error = " << error << endl;
    }

    return 0;
}
